feat(search-with-typeahead): introducing component

IgnacioBecerra opened the pull request:
### Related Ticket(s)
#6143 
#6144

### Description
The current Masthead uses Redux actions in order to incorporate the SearchAPI functionality into `dds-masthead-search`. This includes having a container, composite, and a connect component to render the Masthead Search component, as well as including these actions in the Masthead container and composite components. 

This PR extracts all the functionality from `<dds-masthead-search>` and ensures all the Search API and suggestions functionality can work on its own, meaning that we now effectively have a Search component with Typeahead functionality. 

This PR also replaces the aforementioned Masthead Search components within the actual Masthead. 

### Changelog

**New**

- extracted DDSMastheadSearch functionality into its own component DDSSearchWithTypeahead
- DDSSearchWIthTypeahead React wrapper

**Changed**

- replaced all the DDSMastheadSearch components with the new DDSSearchWithTypeahead component from the Masthead
- refactored some of the search styles

**Removed**

- DDSMastheadSearch, Connect, Container, Composite files

<!-- React and Web Component deploy previews are enabled by default. -->
<!-- To enable additional available deploy previews, apply the following -->
<!-- labels for the corresponding package: -->
<!-- *** "package: services": Services -->
<!-- *** "package: utilities": Utilities -->
<!-- *** "package: styles": Carbon Expressive -->
<!-- *** "RTL": React / Web Components (RTL) -->
<!-- *** "feature flag": React / Web Components (experimental) -->

jeffchew commented:
Thank you @IgnacioBecerra !

I think something happened with the search icon alignment: 

![image](https://user-images.githubusercontent.com/1641214/127360180-961fca6f-e10c-4a5e-bb6f-9cbb76b02be5.png)

jeffchew commented:
@IgnacioBecerra can you add a cdn.html example?

![image](https://user-images.githubusercontent.com/1641214/127360391-cfa575c8-28a0-4538-ab67-de2613976803.png)

jeffchew commented on packages/web-components/src/components/search-with-typeahead/__stories__/README.stories.mdx:
```diff
@@ -0,0 +1,60 @@
+
+Here's a quick example to get you started.
+
+## Example using imports
```
This is missing the CDN documentation.

jeffchew commented on packages/web-components/src/components/search-with-typeahead/__stories__/README.stories.react.mdx:
```diff
@@ -0,0 +1,51 @@
+import contributing from '../../../../../../docs/contributing-license.md';
+import { PropTypesRef } from '@carbon/ibmdotcom-web-components/es/components-react/search-with-typeahead/search-with-typeahead';
+
+# Tag Group
```
Tag Group 😄

jeffchew commented on packages/web-components/src/components/search-with-typeahead/__stories__/README.stories.react.mdx:
```diff
@@ -0,0 +1,51 @@
+
+| Web Components selector    | Compatibility selector | Description |
+| -------------------------- | ---------------------- | ----------- |
+| `<DDSSearchWithTypeahead>` | N/A                    | Component   |
```
There's no compatibility selector that can be used?

jeffchew commented:
Hey @IgnacioBecerra , the search opened onload seems off:

![image](https://user-images.githubusercontent.com/1641214/127505227-420b9285-5d05-4b6a-bd8e-1d5be948d9ec.png)

annawen1 commented on packages/web-components/src/components/masthead/__stories__/masthead.stories.ts:
```diff
@@ -92,7 +91,7 @@ export const WithCustomNavigation = ({ parameters }) => {
       ${styles}
     </style>
-    <dds-masthead-container-search-only
+    <dds-masthead-container
```
@IgnacioBecerra if we remove `dds-masthead-container-search-only` and set it to just `dds-masthead-container`, the Custom Navigation menu items get overwritten from the data in the translation files. Let's switch back to `dds-masthead-composite`?

shixiedesign commented:
Hey @IgnacioBecerra ! Looking pretty good. I'm still checking with designers to see if there are some parts of the spec that's out of date, but here are a few immediate issues to look into: 

1. drop shadow is too strong. Expecting `X:0; Y:0; Blur: 6px; Alpha: 0.3; Color: black`
 
![image](https://user-images.githubusercontent.com/15144993/128058022-e84d6866-917c-40a2-9b9d-1eed650b50db.png)

2. Focus state is a bit unusual. So seems like it's possible to have multiple focus states. Usually I'd expect only one. When typeahead is open, the focus should go into the dropdown, and on tab, if it leaves dropdown, I expect the dropdown option's blue focus outline to disappear. Currently, it doesn't, so resulting in 2 focused elements: 

![Aug-03-2021 13-24-30](https://user-images.githubusercontent.com/15144993/128059351-a3af2e9f-c9d9-418f-b0ca-d65e626bc42e.gif)

**Expecting:** 

![Aug-03-2021 13-28-56](https://user-images.githubusercontent.com/15144993/128059906-722631c3-6b9a-418a-90e3-c4c1634fd6fa.gif)

IgnacioBecerra commented:
@shixiedesign Thanks for the feedback! Since the #6143 issue was mainly about extracting the Masthead Search into its own component, I'm not including the Leadspace Search variation in this PR, but rather in the other PR that you mentioned to have everything pertaining to that component in its own.

Latest commit addresses your comments!

shixiedesign reviewed (approved):
Thanks! Looks good to me 👍

jeffchew commented:
@IgnacioBecerra can you add cypress tests in `/packages/web-components/tests/e2e/__tests__`?

jeffchew commented:
@IgnacioBecerra the CDN example looks off as well:

![image](https://user-images.githubusercontent.com/1641214/128190515-c1c3fcb8-8d97-4bb2-82a5-70f1d9e481de.png)

jeffchew reviewed (approved):
Thank you @IgnacioBecerra !
